airflow/hooks/hdfs_hook.py

```diff
@@ -17,85 +17,95 @@
 
-class HDFSHookException(AirflowException):
-    pass
+    :param str hdfs_conn_id: Connection ID to fetch parameters from.
+    :param bool autoconf: Whether to use autoconfig to discover
+        configuration options from the hdfs XML configuration files.
+    """
 
+    def __init__(self, hdfs_conn_id=None, autoconf=True):
+        super().__init__(None)
 
-class HDFSHook(BaseHook):
-    """
-    Interact with HDFS. This class is a wrapper around the snakebite library.
-
-    :param hdfs_conn_id: Connection id to fetch connection info
-    :type conn_id: string
-    :param proxy_user: effective user for HDFS operations
-    :type proxy_user: string
-    :param autoconfig: use snakebite's automatically configured client
-    :type autoconfig: bool
-    """
-    def __init__(self, hdfs_conn_id='hdfs_default', proxy_user=None,
-                 autoconfig=False):
-        if not snakebite_imported:
-            raise ImportError(
-                'This HDFSHook implementation requires snakebite, but '
-                'snakebite is not compatible with Python 3 '
-                '(as of August 2015). Please use Python 2 if you require '
-                'this hook  -- or help by submitting a PR!')
         self.hdfs_conn_id = hdfs_conn_id
-        self.proxy_user = proxy_user
-        self.autoconfig = autoconfig
+        self._autoconf = autoconf
+
+        self._conn = None
 
     def get_conn(self):
-        """
-        Returns a snakebite HDFSClient object.
-        """
-        # When using HAClient, proxy_user must be the same, so is ok to always
-        # take the first.
-        effective_user = self.proxy_user
-        autoconfig = self.autoconfig
-        use_sasl = configuration.conf.get('core', 'security') == 'kerberos'
-
-        try:
-            connections = self.get_connections(self.hdfs_conn_id)
-
-            if not effective_user:
-                effective_user = connections[0].login
-            if not autoconfig:
-                autoconfig = connections[0].extra_dejson.get('autoconfig',
-                                                             False)
-            hdfs_namenode_principal = connections[0].extra_dejson.get(
-                'hdfs_namenode_principal')
-        except AirflowException:
-            if not autoconfig:
-                raise
-
-        if autoconfig:
-            # will read config info from $HADOOP_HOME conf files
-            client = AutoConfigClient(effective_user=effective_user,
-                                      use_sasl=use_sasl)
-        elif len(connections) == 1:
-            client = Client(connections[0].host, connections[0].port,
-                            effective_user=effective_user, use_sasl=use_sasl,
-                            hdfs_namenode_principal=hdfs_namenode_principal)
-        elif len(connections) > 1:
-            nn = [Namenode(conn.host, conn.port) for conn in connections]
-            client = HAClient(nn, effective_user=effective_user,
```

Comment:
How is the high availability case handled in the new version?

`hdfs3` will read from the configuration files the HA settings (see [docs]), but if we're getting the configuration from the connection, we need to at least specify how to do so in the extra section.

I was thinking that we could specify a `ha` key in the extra section, so that extra looks like

```python
{
  'ha': {
    host = "nameservice1"
    conf = {
        "dfs.nameservices": "nameservice1",
        "dfs.ha.namenodes.nameservice1": "namenode113,namenode188",
        "dfs.namenode.rpc-address.nameservice1.namenode113": "hostname_of_server1:8020",
        "dfs.namenode.rpc-address.nameservice1.namenode188": "hostname_of_server2:8020",
        "dfs.namenode.http-address.nameservice1.namenode188": "hostname_of_server1:50070",
        "dfs.namenode.http-address.nameservice1.namenode188": "hostname_of_server2:50070"
    }
} # number of braces might be wrong here :)
```

and then in the code we could do 

```python
ha = params.extra_dejson.get('ha', {})
if ha:
    pars.update(ha.get('conf'))

...
self._conn = hdfs3.HDFileSystem(
    host=params.host or ha.get('host') or MyNone
...
)
```

What do you think?

[docs]: https://hdfs3.readthedocs.io/en/latest/hdfs.html?highlight=kerberos#high-availability-mode

Reply:
The way other services support HA via connections is to have multiple rows in the Connections table with different host names but the same conn_id.

(Sorry, quick note, will try to expand on this later)

Reply:
Also known as the poor mans load balancing. This will not really work since there is no fallback, but it will pick a random connection, and then you need to pray that one is up.

Reply:
True.

Could we add a new method to Connection base/hook to get all connections with the given ID, and use that here?

Reply:
Yes, that would work.

Apart from the HA stuff, I think we should also drop hdfs3 and go for PyArrow: https://arrow.apache.org/docs/python/filesystems.html